File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.py
dices = []
result = []
maxx = 0    
visited = []

def makeAllCases(cnt, choices, total, sumList):
    if cnt == len(choices):
        sumList.append(total)
        return
        
    for i in range(6):
        makeAllCases(cnt+1, choices, total + choices[cnt][i], sumList)
 
def combination(A, B, diceA, diceB):
    global result
    global maxx
    
    total = 0
    sumA = []
    sumB = []
    
    makeAllCases(0, A, 0, sumA)
    makeAllCases(0, B, 0, sumB)
    
    sumA.sort()
    sumB.sort()
    
    for i in sumA:
        start = 0
        end = len(sumA) - 1
        
        while start <= end:
            mid = (start + end) // 2

            if sumB[mid] >= i:
                end = mid - 1
            else:
                start = mid + 1
        
        total += end
        
    if total > maxx:
        maxx = total 
        result = diceA
    
def dfs(start, cnt, diceCnt):
    if cnt == diceCnt//2:
        A = []
        diceA = []
        B = []
        diceB = []
    
        for i in range(diceCnt):
            if visited[i]:
                A.append(dices[i])
                diceA.append(i+1)
            else:
                B.append(dices[i])
                diceB.append(i+1)
                
        combination(A, B, diceA, diceB)
        return
    
    for i in range(start, diceCnt):  
        if visited[i]:
            continue
            
        visited[i] = 1
        dfs(i, cnt+1, diceCnt)
        visited[i] = 0
        
def solution(dice):
    global visited
    global dices
    global result
    
    answer = []
    dices = dice
    visited = [0] * len(dice)
    
    dfs(0, 0, len(dice))
    
    return result

File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.java
import java.util.*;

class Solution {
    static List<int[]> dices = new ArrayList<>();
    static List<Integer> result = new ArrayList<>();
    static int maxx = 0;
    static boolean[] visited;

    public static void makeAllCases(int cnt, int[][] choices, int total, List<Integer> sumList) {
        if (cnt == choices.length) {
            sumList.add(total);
            return;
        }

        for (int i = 0; i < 6; i++) {
            makeAllCases(cnt + 1, choices, total + choices[cnt][i], sumList);
        }
    }

    public static void combination(int[][] A, int[][] B, List<Integer> diceA, List<Integer> diceB) {
        int total = 0;
        List<Integer> sumA = new ArrayList<>();
        List<Integer> sumB = new ArrayList<>();

        makeAllCases(0, A, 0, sumA);
        makeAllCases(0, B, 0, sumB);

        Collections.sort(sumA);
        Collections.sort(sumB);

        for (int i : sumA) {
            int start = 0;
            int end = sumB.size() - 1;

            while (start <= end) {
                int mid = (start + end) / 2;

                if (sumB.get(mid) >= i) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            }

            total += end;
        }

        if (total > maxx) {
            maxx = total;
            result = new ArrayList<>(diceA);
        }
    }

    public static void dfs(int start, int cnt, int diceCnt) {
        if (cnt == diceCnt / 2) {
            List<int[]> A = new ArrayList<>();
            List<Integer> diceA = new ArrayList<>();
            List<int[]> B = new ArrayList<>();
            List<Integer> diceB = new ArrayList<>();

            for (int i = 0; i < diceCnt; i++) {
                if (visited[i]) {
                    A.add(dices.get(i));
                    diceA.add(i + 1);
                } else {
                    B.add(dices.get(i));
                    diceB.add(i + 1);
                }
            }

            combination(A.toArray(new int[0][]), B.toArray(new int[0][]), diceA, diceB);
            return;
        }

        for (int i = start; i < diceCnt; i++) {
            if (visited[i]) continue;

            visited[i] = true;
            dfs(i + 1, cnt + 1, diceCnt);
            visited[i] = false;
        }
    }

    public static List<Integer> solution(int[][] dice) {
        dices = Arrays.asList(dice);
        visited = new boolean[dice.length];
        dfs(0, 0, dice.length);
        return result;
    }

}